src/ui/public/modals/__tests__/confirm_modal.js

```diff
@@ -0,0 +1,186 @@
+    });
+
+    it('when showClose is on but title is not given', function () {
+      const options = { customConfirmButton: 'b', onConfirm: _.noop, showClose: true };
+      expect(() => confirmModal('hi', options)).to.throwError();
+    });
+  });
+
+  it('shows the message', function () {
+    const myMessage = 'Hi, how are you?';
+    confirmModal(myMessage, { confirmButtonText: 'GREAT!', onConfirm: _.noop });
+
+    $rootScope.$digest();
+    const message = findByDataTestSubj('confirmModalBodyText')[0].innerText;
+    expect(message).to.equal(myMessage);
+  });
+
+  describe('shows custom text', function () {
+    const confirmModalOptions = {
+      confirmButtonText: 'Troodon',
+      cancelButtonText: 'Dilophosaurus',
+      title: 'Dinosaurs',
+      onConfirm: _.noop
+    };
+
+    it('for confirm button', () => {
+      confirmModal('What\'s your favorite dinosaur?', confirmModalOptions);
+      $rootScope.$digest();
+      const confirmButtonText = findByDataTestSubj('confirmModalConfirmButton')[0].innerText;
+      expect(confirmButtonText).to.equal('Troodon');
+    });
+
+    it('for cancel button', () => {
+      confirmModal('What\'s your favorite dinosaur?', confirmModalOptions);
+      $rootScope.$digest();
+      const cancelButtonText = findByDataTestSubj('confirmModalCancelButton')[0].innerText;
+      expect(cancelButtonText).to.equal('Dilophosaurus');
+    });
+
+    it('for title text', () => {
+      confirmModal('What\'s your favorite dinosaur?', confirmModalOptions);
+      $rootScope.$digest();
+      const titleText = findByDataTestSubj('confirmModalTitleText')[0].innerText;
+      expect(titleText).to.equal('Dinosaurs');
+    });
+  });
+
+  describe('x icon', function () {
+    it('is visible when showClose is true', function () {
+      const confirmModalOptions = {
+        confirmButtonText: 'bye',
+        onConfirm: _.noop,
+        showClose: true,
+        title: 'hi'
+      };
+      confirmModal('hi', confirmModalOptions);
+
+      $rootScope.$digest();
+      const xIcon = findByDataTestSubj('confirmModalCloseButton');
+      expect(xIcon.length).to.be(1);
+    });
+
+    it('is not visible when showClose is false', function () {
+      const confirmModalOptions = {
+        confirmButtonText: 'bye',
+        onConfirm: _.noop,
+        title: 'hi',
+        showClose: false
+      };
+      confirmModal('hi', confirmModalOptions);
+
+      $rootScope.$digest();
+      const xIcon = findByDataTestSubj('confirmModalCloseButton');
+      expect(xIcon.length).to.be(0);
+    });
+  });
+
+  describe('callbacks are called:', function () {
+    const confirmCallback = sinon.spy();
+    const closeCallback = sinon.spy();
+    const cancelCallback = sinon.spy();
+
+    const confirmModalOptions = {
+      confirmButtonText: 'bye',
+      onConfirm: confirmCallback,
+      onCancel: cancelCallback,
+      onClose: closeCallback,
```

Comment:
What would you propose instead?  I can rename confirm => custom and rename all the buttons from cancel => secondary action and confirm => primary action? 

I don't want to introduce a slew of extremely similar code just so I can can pass a different onClose than onCancel action, and that is the only difference between the modal that I need and the confirmModal that you would prefer.

Would you be content if I opened an issue for to explore this later?  It feels like diminishing returns at this point.  IMO it is not far fetched to allow a confirmation modal a custom onClose as a specialized option (not the default).  Just because the native confirmation window works like that doesn't mean this has to.  We allow them to set the cancel button text, so the user might not be answering 'no' or 'cancel', but something else.

Other "Confirmation" dialogs offer more than just yes/no as well:
![customconfirm1](https://cloud.githubusercontent.com/assets/16563603/22382532/0c7724ba-e494-11e6-8e6e-84069d1bd628.gif)

![customconfirm2](https://cloud.githubusercontent.com/assets/16563603/22382414/98bcdac4-e493-11e6-9ccf-44f2fbd7e06a.gif)

And jquery-confirm also doesn't make the close function match the onCancel function (at least I think that's how it works by reading the functions - looks like close just closes without calling onCancel) http://craftpip.github.io/jquery-confirm/v1.7.8/

I'm fine with improving this at some point, but I don't think it's the highest priority.  Seems like something we could tackle when we address how to handle even more customized modals (e.g. passing in templates, etc).  So unless there is a quick-ish solution that doesn't cause a bunch of duplicate code, I'd prefer to punt.